ISSUE-785 Revisit manual service registerer

HeartSaVioR opened the pull request:
This closes #785 

* Define required and optional fields instead of uploading config file
  * we missed that config file doesn't have all fields and streamline can't infer default value
* Introduce ComponentProcess to break down Component and Process
  * hosts / port / protocol are moved from Component to ComponentProcess
  * WARN: table schema changed

@hmcl 
I modified lots of metadata codes because of ComponentProcess. 
Could you help me reviewing the code and see if I'm breaking anything? Thanks in advance.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/hortonworks/streamline/792)
<!-- Reviewable:end -->

HeartSaVioR commented:
Manually tested with unsecured cluster: both manual / imported cluster with Kafka only topology.

HeartSaVioR commented:
This patch also fixes test failing on RestIntegrationTest.testAllResources. I disabled topology REST API test because it now depends on corresponding complicated entities (Namespace, Service, Component, ComponentProcess, ServiceConfiguration, Mapping).

HeartSaVioR commented:
This is quite a huge patch but I would be really happy if someone can take a stab with reviewing. I'll address conflicts after passing review process.

harshach commented:
@hmcl can you review this.

hmcl commented on streams/cluster/src/test/java/com/hortonworks/streamline/streams/cluster/register/impl/KafkaServiceRegistrarTest.java:
```diff
@@ -51,47 +54,70 @@ public void testRegister() throws Exception {
+        assertNotNull(broker);
+
+        Collection<ComponentProcess> brokerProcesses = environmentService.listComponentProcessesInComponent(broker.getId());
+        assertEquals(4, brokerProcesses.size());
```
why 4? Is it NUM_LISTENERS?  can you put a constant describing what this number is?

HeartSaVioR replied:
Yes number of component processes for broker. I'll put a comment.

hmcl commented on streams/cluster/src/test/java/com/hortonworks/streamline/streams/cluster/register/impl/StormServiceRegistrarTest.java:
```diff
@@ -57,38 +61,62 @@ public void testRegister() throws Exception {
+
+        Collection<ComponentProcess> nimbusProcesses = environmentService.listComponentProcessesInComponent(nimbus.getId());
+        List<String> hosts = nimbusProcesses.stream().map(ComponentProcess::getHost).collect(Collectors.toList());
+        assertEquals(Sets.newHashSet("storm-1", "storm-2"), new HashSet<>(hosts));
```
Is this always true, or does it depend on the order of the sets? If so it could cause sporadic test failures.

HeartSaVioR replied:
Java `Set` instance declares this clearly. No need to worry about order.

```
    /**
     * Compares the specified object with this set for equality.  Returns
     * <tt>true</tt> if the specified object is also a set, the two sets
     * have the same size, and every member of the specified set is
     * contained in this set (or equivalently, every member of this set is
     * contained in the specified set).  This definition ensures that the
     * equals method works properly across different implementations of the
     * set interface.
     *
     * @param o object to be compared for equality with this set
     * @return <tt>true</tt> if the specified object is equal to this set
     */
    boolean equals(Object o);
```

hmcl replied:
Sure. I was just making sure :)

hmcl commented on streams/cluster/src/test/java/com/hortonworks/streamline/streams/cluster/register/impl/StormServiceRegistrarTest.java:
```diff
@@ -57,38 +61,62 @@ public void testRegister() throws Exception {
+        List<String> hosts = nimbusProcesses.stream().map(ComponentProcess::getHost).collect(Collectors.toList());
+        assertEquals(Sets.newHashSet("storm-1", "storm-2"), new HashSet<>(hosts));
+        List<Integer> ports = nimbusProcesses.stream().map(ComponentProcess::getPort).collect(Collectors.toList());
+        assertEquals(Sets.newHashSet(6627, 6627), new HashSet<>(ports));
```
Is this always true, or does it depend on the order of the sets? If so it could cause sporadic test failures.

HeartSaVioR replied:
Same here.

hmcl commented on streams/cluster/src/test/java/com/hortonworks/streamline/streams/cluster/register/impl/KafkaServiceRegistrarTest.java:
```diff
@@ -101,15 +127,17 @@ public void testRegister_component_kafka_broker_notPresent() throws Exception {
     }
 
     @Test
```
NIT: you could use an annotation saying that this test is expected to throw IAE

HeartSaVioR replied:
Same here.

hmcl commented on streams/cluster/src/test/java/com/hortonworks/streamline/streams/cluster/register/impl/KafkaServiceRegistrarTest.java:
```diff
@@ -51,47 +54,70 @@ public void testRegister() throws Exception {
+        assertFalse(serverPropertiesConfMap.containsKey(KafkaServiceRegistrar.PARAM_SECURITY_INTER_BROKER_PROTOCOL));
     }
 
     @Test
```
NIT: you could use an annotation saying that this test is expected to throw IAE

HeartSaVioR replied:
I intentionally catch the exception to check service is not added.

hmcl replied:
Makes sense. Thanks.

hmcl commented on streams/cluster/src/test/java/com/hortonworks/streamline/streams/cluster/register/impl/StormServiceRegistrarTest.java:
```diff
@@ -97,27 +125,10 @@ public void testRegisterWithoutNimbusPrincipal() throws Exception {
+        assertFalse(confMap.containsKey(StormServiceRegistrar.PARAM_NIMBUS_PRINCIPAL_NAME));
     }
 
     @Test
```
NIT: you could use an annotation saying that this test is expected to throw IAE

HeartSaVioR replied:
Same here.

hmcl commented on streams/cluster/src/test/java/com/hortonworks/streamline/streams/cluster/register/impl/StormServiceRegistrarTest.java:
```diff
@@ -145,11 +157,12 @@ public void testRegister_component_nimbus_notPresent() throws Exception {
         StormServiceRegistrar registrar = initializeServiceRegistrar();
 
-        try (InputStream is = getClass().getClassLoader().getResourceAsStream(STORM_YAML_FILE_PATH)) {
+        try {
```
NIT: you could use an annotation saying that this test is expected to throw IAE

HeartSaVioR replied:
Same here.

hmcl commented on streams/cluster/src/test/java/com/hortonworks/streamline/streams/cluster/register/impl/ZookeeperServiceRegistrarTest.java:
```diff
@@ -50,71 +55,40 @@ public void testRegister() throws Exception {
-        }
-    }
+        Collection<ComponentProcess> zkServerProcesses = environmentService.listComponentProcessesInComponent(zkServer.getId());
+        assertEquals(Sets.newHashSet("zookeeper-1", "zookeeper-2"),
```
Is this always true, or does it depend on the order of the sets? If so it could cause sporadic test failures.

HeartSaVioR replied:
Same here.

hmcl commented on streams/cluster/src/test/java/com/hortonworks/streamline/streams/cluster/service/metadata/KafkaMetadataServiceTest.java:
```diff
@@ -118,30 +129,28 @@ public void test_KafkaZkConnection_createPath() throws Exception {
 
     @Test
     public void test_getBrokerInfoFromZk() throws Exception {
+        List<ComponentProcess> brokerProcesses = getBrokerComponentProcesses();
```
redundant var

hmcl commented on streams/cluster/src/test/java/com/hortonworks/streamline/streams/cluster/service/metadata/StormMetadataServiceTest.java:
```diff
@@ -85,6 +92,12 @@ public void getTopologies() throws Exception {
         }};
 
+        List<ComponentProcess> nimbusProcesses = buildNimbusComponentProcesses();
+        List<ComponentProcess> uiProcesses = buildUiComponentProcesses();
```
redundant vars

HeartSaVioR commented on bootstrap/services/storm-bundle.json:
```diff
@@ -4,33 +4,62 @@
   "serviceUISpecification": {
     "fields": [
       {
+        "uiName": "Nimbuses hostnames",
```
Storm uses `Nimbuses` and `Nimbodes` to represent plural form of Nimbus.
I'm not clear which word Nathan referred, but at least one of word `nimbus` has `nimbuses` as plural form.
https://www.merriam-webster.com/dictionary/nimbus

hmcl replied:
I learned something :). Thanks.

HeartSaVioR commented on bootstrap/services/storm-bundle.json:
```diff
@@ -4,33 +4,62 @@
+        "uiName": "Principal to local plug-in",
+        "fieldName": "storm.principal.tolocal",
+        "isOptional": true,
+        "tooltip": "The plugin that will convert a principal to a local user",
```
I just copied explanation from current Storm Config doc.
https://github.com/apache/storm/blob/master/storm-client/src/jvm/org/apache/storm/Config.java#L1267-L1271

If we can find better description, we can apply it to both side.

hmcl replied:
same comment

HeartSaVioR commented on bootstrap/services/storm-bundle.json:
```diff
@@ -4,33 +4,62 @@
-        "isOptional": false,
-        "tooltip": "Upload storm.yaml file used by Storm cluster",
-        "type": "file"
+        "uiName": "Principal to local plug-in",
```
Storm clearly states that it is a plug-in to convert principal to local user.
https://github.com/apache/storm/blob/master/storm-client/src/jvm/org/apache/storm/Config.java#L1267-L1271

If we can find better description, we can apply it to both side.

hmcl replied:
OK. I think that this will not be an easy to process info for the user, but we can work on that later. Perhaps fix it in both places as you say if we come up with something better.

HeartSaVioR commented on bootstrap/services/storm-bundle.json:
```diff
@@ -4,33 +4,62 @@
+        "type": "string"
+      },
+      {
+        "uiName": "Nimbus Principal",
```
Kerberos looks better.

HeartSaVioR commented on streams/actions/src/main/java/com/hortonworks/streamline/streams/actions/container/TopologyActionsContainer.java:
```diff
@@ -101,15 +104,30 @@ private TopologyActions initTopologyActions(Map<String, Object> conf, String cla
-        String uiHost = uiServer.getHosts().get(0);
-        Integer uiPort = uiServer.getPort();
+
+        Collection<ComponentProcess> uiServerProcesses = environmentService.listComponentProcessesInComponent(uiServer.getId());
```
We extract the first one and name `uiServerProcess` so it will conflict.

HeartSaVioR commented on streams/actions/src/main/java/com/hortonworks/streamline/streams/actions/container/TopologyActionsContainer.java:
```diff
@@ -101,15 +104,30 @@ private TopologyActions initTopologyActions(Map<String, Object> conf, String cla
+        }
+
+        ComponentProcess uiServerProcess = uiServerProcesses.iterator().next();
+        String uiHost = uiServerProcess.getHost();
```
Lots of local var can be final. Let's adopt style guide and integrate checkstyle to the build phase if we really want to handle style issue.

HeartSaVioR commented on streams/cluster/src/main/java/com/hortonworks/streamline/streams/cluster/register/impl/AbstractServiceRegistrar.java:
```diff
@@ -117,8 +126,26 @@ public Service register(Cluster cluster, Config config, List<ConfigFileInfo> con
         return service;
     }
 
+    protected boolean validateComponentProcesses(List<ComponentProcess> componentProcesses) {
```
plural form and List look like duplicated. `isComponentProcessesValid` looks OK.

HeartSaVioR commented on streams/cluster/src/main/java/com/hortonworks/streamline/streams/cluster/register/impl/AbstractServiceRegistrar.java:
```diff
@@ -117,8 +126,26 @@ public Service register(Cluster cluster, Config config, List<ConfigFileInfo> con
+                        componentProcess.getPort() != null);
+    }
 
+    protected boolean validateComponentProcessesWithProtocolRequired(List<ComponentProcess> componentProcesses) {
```
Same here: `isComponentProcessesWithProtocolRequiredValid` looks OK.

HeartSaVioR commented on streams/cluster/src/main/java/com/hortonworks/streamline/streams/cluster/register/impl/KafkaServiceRegistrar.java:
```diff
@@ -15,67 +15,96 @@
+import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
 
```
There's no HDFS_SITE, and no CORE_SITE.

HeartSaVioR commented on streams/cluster/src/main/java/com/hortonworks/streamline/streams/cluster/register/impl/KafkaServiceRegistrar.java:
```diff
@@ -86,23 +115,68 @@ protected boolean validateServiceConfiguationsAsFlattenedMap(Map<String, String>
+        Map<String, String> confMap = new HashMap<>();
+
+        try {
+            String json = objectMapper.writeValueAsString(confMap);
```
Yes some configurations will be filled while importing from Ambari, but these are actually not needed so just putting empty JSON.

HeartSaVioR commented on streams/cluster/src/main/java/com/hortonworks/streamline/streams/cluster/register/impl/StormServiceRegistrar.java:
```diff
@@ -37,69 +39,65 @@
-    public static final String PARAM_NIMBUS_PRINCIPAL_NAME = "nimbusPrincipalName";
-    public static final String STORM_ENV = ServiceConfigurations.STORM.getConfNames()[1];
+
+    public static final String PARAM_NIMBUS_SEEDS = "nimbus.seeds";
```
That represents config key in UI. If you think config key is better than param, we can change it.

hmcl replied:
Oh, I see. Would UI_PARAM? UI_SET_PARAM?  UI_DEFINED_PARAM be more suiting?

HeartSaVioR commented on streams/cluster/src/main/java/com/hortonworks/streamline/streams/cluster/register/impl/StormServiceRegistrar.java:
```diff
@@ -117,42 +115,124 @@ protected boolean validateServiceConfiguationsAsFlattenedMap(Map<String, String>
+    private Number readNumberFromConfig(Config config, String parameterName) {
+        Number paramValue;
+        try {
+            paramValue = config.getAny(parameterName);
```
As you are seeing that compiler doesn't complain, no need to cast.

HeartSaVioR replied:
Ah OK you meant to reduce lines.

HeartSaVioR commented on streams/cluster/src/main/java/com/hortonworks/streamline/streams/cluster/service/EnvironmentService.java:
```diff
@@ -104,16 +106,24 @@ public Service initializeService(Cluster cluster, String serviceName) {
         return component;
     }
 
+    public ComponentProcess initializeComponentProcess(Component component, String host) {
```
Either is fine so just fixed.

HeartSaVioR commented on streams/cluster/src/main/java/com/hortonworks/streamline/streams/cluster/service/metadata/json/KafkaBrokerListeners.java:
```diff
@@ -55,7 +65,19 @@ public ListenersPropParsed(List<ListenersPropEntry> parsedProps) {
         public ListenersPropParsed(ServiceConfiguration config) {
             try {
-                final Map<String, String> configMap = config.getConfigurationMap();
+                Map<String, String> confMap = config.getConfigurationMap();
```
config.getConfigurationMap() throws IOException so we should add throw to method signature (bad to do that since only this constructor will throw IOE). Let's keep it as it is.

HeartSaVioR commented:
@hmcl 
Thanks for the detailed review. I addressed review comments. 
Btw, I guess you're still in reviewing. If then please let me know if you finish your review. Thanks again!

HeartSaVioR commented:
Filed #875 . Thanks for the review @hmcl. I'll merge.

HeartSaVioR commented:
Merged via 0b9279e873c6d6d21e20121613959d2799be2d8d
